Add pypardiso recipe

cmutel opened the pull request:
Library for sparse matrix math using the pardiso solver in the Intel MKL library.

Recipe builds and tests successfully on my machine.

Source is **not** from original source, but from a fork which implemented complex matrices. My intention is to expand on this library eventually to support newer versions of pardiso.

I confirm that I am willing to maintain this package.

Pinging @conda-forge/help-python :)

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/pypardiso```) and found some lint.

Here's what I've got...


For **recipes/pypardiso**:

 * ``requirements: run: libblas=*=*mkl`` must contain a space between the name and the pin, i.e. ``libblas =*=*mkl``

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pypardiso```) and found it was in an excellent condition.

isuruf commented:
Can you use the original source and add the changes in the fork as patches?

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/pypardiso```) and found some lint.

Here's what I've got...


For **recipes/pypardiso**:

 * Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form. See lines [26, 27, 35, 36]

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pypardiso```) and found it was in an excellent condition.

cmutel commented:
@isuruf I switched to the original source, and after some recipe debugging, everything builds correctly.

cmutel commented:
Pinging @conda-forge/help-python, I think this is ready for review and possible approval.

haasad commented:
> GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.

I confirm that I want to be listed as a maintainer

cmutel commented:
@conda-forge/staged-recipes @isuruf Would be nice to have this reviewed and merged :)

chrisburr commented on recipes/pypardiso/meta.yaml:
```diff
@@ -0,0 +1,60 @@
+  version: {{ version }}
+
+source:
+  url: https://github.com/haasad/PyPardisoProject/archive/refs/tags/v0.2.2.tar.gz
```
```suggestion
  url: https://github.com/haasad/PyPardisoProject/archive/refs/tags/v{{ version }}.tar.gz
```

chrisburr commented on recipes/pypardiso/meta.yaml:
```diff
@@ -0,0 +1,60 @@
+  skip: True  # [py<35]
+
+requirements:
+  build:
+    - python
+    - setuptools
```
```suggestion
```

chrisburr commented on recipes/pypardiso/meta.yaml:
```diff
@@ -0,0 +1,60 @@
+    - setuptools
+    - numpy
+    - pip
+    - "libblas =*=*mkl"  # [win]
+    - mkl 2020.*  # [win]
```
```suggestion
```

cmutel replied:
This is necessary for the builds to work properly. See https://github.com/conda-forge/staged-recipes/pull/15106/commits/1ab96ed6cec87034da962e013767caa5eb2d6a03 and failures on previous versions.

chrisburr commented on recipes/pypardiso/meta.yaml:
```diff
@@ -0,0 +1,60 @@
+    - {{ pin_compatible('numpy') }}
+    - "libblas =*=*mkl"
+    - scipy
+    - mkl  # [not win]
+    - mkl 2020.*  # [win]
```
Why does this need to be constrained on Windows?
```suggestion
    - mkl
```

cmutel replied:
The latest releases of MKL changed the name of the linked library, at least on Windows, so this is necessary to get the Windows version to build correctly.

isuruf replied:
I think you need `mkl-devel` and it should work

cmutel replied:
Yes; I updated the recipe.

chrisburr commented on recipes/pypardiso/meta.yaml:
```diff
@@ -0,0 +1,60 @@
+build:
+  number: 0
+  script: "{{ PYTHON }} -m pip install . -vv"
+  skip: True  # [py<35]
```
If my other commends are valid this looks like it can be `noarch: python`:

```suggestion
  noarch: python
```

cmutel replied:
If there a cost to build a separate package for each version? We have already lost a decent amount of time debugging specific problems with versions and OSes, my preference would be to use the well-tested conda-forge architecture to build known good packages for each version and OS combination automatically.

cmutel replied:
Seems to build find with `noarch`

chrisburr commented on recipes/pypardiso/meta.yaml:
```diff
@@ -0,0 +1,60 @@
+    - mkl 2020.*  # [win]
+  run:
+    - python
+    - {{ pin_compatible('numpy') }}
```
```suggestion
    - numpy
```

chrisburr commented on recipes/pypardiso/meta.yaml:
```diff
@@ -0,0 +1,60 @@
+  host:
+    - python
+    - setuptools
+    - numpy
```
```suggestion
```

cmutel replied:
Sorry, the distinction between host and build is still a bit fuzzy for me. I am worried that this will somehow break things, but we can commit it and try...

chrisburr commented on recipes/pypardiso/meta.yaml:
```diff
@@ -0,0 +1,60 @@
+    - tests/test_scipy_aliases.py
+    - tests/test_input_b.py
+  commands:
+    - pytest -vv tests/
```
```suggestion
    - pytest -vv tests/
    - pip check
```

chrisburr commented on recipes/pypardiso/meta.yaml:
```diff
@@ -0,0 +1,60 @@
+  imports:
+    - pypardiso
+  requires:
+    - pytest
```
```suggestion
    - pytest
    - pip
```

cmutel commented:
@chrisburr Thanks a lot for your detailed review. I think I have addressed most of your comments, except for one which I don't know the answer to (if there is one answer). I think we can't build a `noarch` version due to the Windows requirement of using an older version of Intel MKL.

ma-sadeghi commented:
Hi @cmutel, is there a reason you've hardcoded MKL 2020 for Windows?

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/pypardiso```) and found some lint.

Here's what I've got...


For **recipes/pypardiso**:

 * noarch: python recipes are required to have a lower bound on the python version. Typically this means putting `python >=3.6` in **both** `host` and `run` but you should check upstream for the package's Python compatibility.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pypardiso```) and found it was in an excellent condition.

cmutel commented:
@ma-sadeghi @isuruf @chrisburr I have updated the recipe to use 0.3.1, which alleviates the need for specific MKL version pinning on Windows, and allows for a `noarch` build.

ma-sadeghi commented:
@cmutel Great, thanks!

ma-sadeghi commented:
@conda-forge/help-python could you please take a look at this PR? Two of our libraries depend on `pypardiso` and are currently broken. Really appreciate your help :)

ma-sadeghi commented:
@cmutel I think if you list PyPI as the source url (instead of GitHub), conda-forge automatically creates a PR whenever a new release is out. Once this PR is merged you can also add an automerge bot to the conda feedstock so PRs get merged automatically if build is successful and all tests pass. This way, you don't really need to do anything for maintaining the conda-forge feedstock for `pypardiso`. That's how we do it for our two libraries `openpnm` and `porespy` and I find it very convenient that I no longer need to worry about updating the conda-forge release every time a new release is out. Anyway, just a suggestion. You just need to update the source url to
```
https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
```

ma-sadeghi commented:
Hi @beckermr! Can you please mark https://anaconda.org/conda-forge/pypardiso as broken? Thanks!

beckermr commented:
I will fix the outputs. Please submit a PR to conda-forge/admin-requests to mark the current builds broken.

beckermr commented:
PR made: https://github.com/conda-forge/feedstock-outputs/pull/18

Once this PR is merge, bump me to merge the one above and you should be good to go.

beckermr commented:
OK. All of the PRs are merged. This PR can be merged when the current reviewers are happy.

cmutel commented:
@ma-sadeghi Thanks Amin. I originally used the Github link because there wasn't a pypi release, but now that there is, of course we should switch.

ma-sadeghi commented on recipes/pypardiso/meta.yaml:
```diff
@@ -0,0 +1,56 @@
+    - tests/test_input_b.py
+  commands:
+    - pytest -vv tests/
+    - pip check
```
@cmutel The builds are failing because of this line, which I think you can get rid of. It basically checks which packages are installed whose dependency is not. I think since `mkl` is installed using `conda`, `pip` fails (I think for `pip check` to succeed, it needs all dependencies installed using `pip`). Anyway, you're already testing `pypardiso` in `pytest -vv tests/`, so if there's a dependency missing, it will be caught during the `import pypardiso`.

cmutel replied:
@conda-forge/help-python Any idea why `pip check` would fail with [missing dependencies](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=339419&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=829) after switching from the Github release to the pypi release? I don't understand hwo things could be [just fine](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=338502&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=780) with everything the same except the location of the downloaded tarball...

isuruf replied:
It's because of `mkl` in https://github.com/haasad/PyPardisoProject/blob/master/setup.py#L7

haasad replied:
Let's use the github source for now in that case, I'll tidy it up for the next version. Probably my lack of experience with PyPI that caused this.

cmutel replied:
@isuruf Yeah, but no :) `mkl` is installed into the test environment in both the [passing](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=338502&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=704) and [failing](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=339419&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=752) tests. `pip` is not used to install anything, it is all conda. And pip could find mkl on its path in the passing tests...

The `pip check` suggestion came from someone who knows what they are doing, so I think it is best practice in conda-forge recipes.

@haasad The pypi upload looks fine to me, I assume you did `python setup.py sdist bdist_wheel upload` or used `twine`, there isn't much you could do wrong.

isuruf replied:
There's a pip package called `mkl` which `pip check` is looking for, but we are installing the conda package. So, the best solution is to remove `mkl` in `setup.py` and keep `mkl-service` which has `mkl` as a dependency.

haasad replied:
@cmutel I released version 0.3.2, where `mkl` is removed as direct dependency in `setup.py`: https://pypi.org/project/pypardiso/

ma-sadeghi replied:
@cmutel Could you please update `meta.yaml` to version 0.3.2 (+ the sha256)? :)

isuruf commented on recipes/pypardiso/meta.yaml:
```diff
@@ -0,0 +1,56 @@
+    - pip check
+
+about:
+  home: https://github.com/cmutel/PyPardisoProject
```
```suggestion
  home: https://github.com/haasad/PyPardisoProject
```

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/pypardiso```) and found it was in an excellent condition.


I do have some suggestions for making it better though...


For **recipes/pypardiso**:

 * A conda package with same name (pardiso4py) already exists.

ma-sadeghi commented:
@beckermr Do you know how long it takes before `pypardiso` is fully deleted from Anaconda.org? 
https://anaconda.org/conda-forge/pypardiso

ma-sadeghi commented:
Hi @isuruf, could you please review this PR one more time? It seems that it's in good shape. Very much appreciate it.

isuruf commented on recipes/pypardiso/meta.yaml:
```diff
@@ -0,0 +1,56 @@
+  run:
+    - python >=3.6
+    - numpy
+    - "libblas =*=*mkl"
```
```suggestion
    - libblas * *mkl
```
